Enhancement: conditional links

Hurleveur opened the pull request:
**Feature:** This PR implements the feature requested by @beccabunny

## Feature Details
Links may have a child node.
They behave exactly as other conditions do, meaning you could also do ands, or etc.
They will get refreshed each time you jump. Eventually we'd get this for all objects.

## UI Screenshots
n/a

## Usage Examples
idk, some weird maze? sometimes showing the path to hidden systems?
One could do this to make a random link that sometimes appear.
`link "Sol"`
`    random < 40`
or if you want it to only be active when you're next to it, for instance
`link "Sol"`
`    has "flagship system: Alpha Centauri"`

## Testing Done
not yet

### Automated Tests Added
Added one to make sure the way back is always guarantied and to test when its 1 and 0 since I cant test random.

## Performance Impact
refreshes all systems on entry of a system

1010todd commented:
What's the syntax? Can we specify chance it will appear?

Hurleveur commented:
`"random link" <system> <chance>` is the syntax, i had written it but github hid it caus of using `<` and `>`

Zitchas commented:
This would be *very* useful for upcoming Ember Waste content.

That being said, I'd prefer if it was strictly the random chance, and no "guaranteed reverse course." Preventing players from getting trapped is the content creator's responsibility, and there's plenty of ways to do so. If they want to make a guaranteed way back, then there's a random chance for the A to B, then a normal link from B  to A, 

Thought: What is the frequency of change? Is it daily? Would it be possible to set a frequency so we could have a link that, say, will stay open for 7 days, then closed for 3, and repeat? Perhaps that'd be better as a new PR.

Anyway, random links will be useful. Thanks for making this!

Hurleveur commented:
becca asked for the garanteed way back I don't have opinion either way
the frequency of the change is daily currently, yeah I could complexify it but that'd be for another PR

Zitchas commented:
Given that we can make one way hyperlinks that don't provide a way back, I don't think having a random hyperlink forcefully establish a way back fits with our existing code/data, and it does actually impairs the functionality.

At the very least, there needs to be a way to explicitly disable said "automatic way back"

Hurleveur commented:
I'll add that per system to make it simple

Zitchas commented:
Thanks! So, something like `"random link" <system> <chance> <supresswayback>`

Hurleveur commented:
Just saw why the test wasnt succeeding...
Everything works but the shuttle is dancing to head back, turning left and right all the time and never jumping... Fixed by making it slower.

Now, what I meant was a per system attribute; `"guaranteed link back"` which I just added. If you write remove "guaranteed way back" then there wont be one.
Do you really need it per link? I think it's best to do it simple like this for now and eventually expand it when a class etc is done for links. Right now it's not worth doing a class just for this though.

warp-core commented on source/Engine.cpp:
```diff
@@ -272,7 +272,7 @@ void Engine::Place()
 	ai.ClearOrders();
 
-	EnterSystem();
+	EnterSystem(const_cast<System &>(*player.GetSystem()));
```
What's the cast for?

Hurleveur replied:
~~ask the compiler~~

I need system and not const system to be able to link and unlink systems since that takes a non cast System *

tehhowch replied:
That cast is a red flag, telling you and others, "this should be done someplace else."

Hurleveur replied:
I'll make GameData do it internally then since it holds the objects.
That said I'm wondering, is it best to search for that system in GameData to have a non cost object or to cast it once in it?

warp-core commented on source/Engine.cpp:
```diff
@@ -1202,7 +1202,7 @@ void Engine::BreakTargeting(const Government *gov)
 
 
-void Engine::EnterSystem()
+void Engine::EnterSystem(System &previousSystem)
```
`const`?

Hurleveur replied:
this part is temporary it wont stay, just wrote it to get this previous system but it'll be provided by playerInfo

that said I need it non const anyway later on so I don't think it'd be worth it

Hurleveur replied:
https://github.com/endless-sky/endless-sky/pull/7974 this will provide the means to get it
but I'll have to const cast it anyway

warp-core commented:
> Thanks! So, something like `"random link" <system> <chance> <supresswayback>` ?

There should also be the option to neither guarantee a link is still present immediately after it is traversed nor that that link will be gone.

> Just saw why the test wasnt succeeding... Everything works but the shuttle is dancing to head back, turning left and right all the time and never jumping... Fixed by making it slower.
> 
> Now, what I meant was a per system attribute; `"guaranteed link back"` which I just added. If you write remove "guaranteed way back" then there wont be one. Do you really need it per link? I think it's best to do it simple like this for now and eventually expand it when a class etc is done for links. Right now it's not worth doing a class just for this though.

The randomness is per link, whether or not the link is maintained to allow a reverse trip, explicitly not present after it is traversed, or otherwise should also be per link.

warp-core commented on source/System.cpp:
```diff
@@ -99,7 +99,7 @@ void System::Load(const DataNode &node, Set<Planet> &planets)
 	// For the following keys, if this data node defines a new value for that
 	// key, the old values should be cleared (unless using the "add" keyword).
-	set<string> shouldOverwrite = {"asteroids", "attributes", "belt", "fleet", "link", "object", "hazard"};
+	set<string> shouldOverwrite = {"asteroids", "attributes", "belt", "fleet", "link", "random link", "object", "hazard"};
```
Can you move `hazard` to between `fleet` and `link` and put `random link` at the end so its alphabetised?

warp-core commented on source/System.cpp:
```diff
@@ -957,6 +987,13 @@ void System::UpdateNeighbors(const Set<System> &systems, double distance)
 	// even if it is farther away than the maximum distance.
 	for(const System *system : links)
 		neighborSet.insert(system);
+	// Even if we are not permanently connected to that system via hyperdrive,
+	// It will still be reachable by Jump Drive.
+	for(auto &link : randomLinks)
+		neighborSet.insert(link.first);
```
Can a "random link" always be traversed by jump drive, even if the link isn't actually present?

Hurleveur replied:
the random link just is a hyperdrive link that is sometimes there and sometimes not
I don't see a reason to prevent JD travel, that'd be done by another functionality

warp-core replied:
So, why are you saying the system will still be reachable by jump drive even if hte random link isn't actually there?

Hurleveur replied:
a jump drive doesnt need a hyperdrive link to reach a system? links always have been hyperdrive links

warp-core replied:
There's still the consideration of distance, which you seem to be ignoring here.

Hurleveur replied:
I don't think so? do you mean that with this you could reach an out of reach system by jumping? I'll test it

Hurleveur commented:
> The randomness is per link, whether or not the link is maintained to allow a reverse trip, explicitly not present after it is traversed, or otherwise should also be per link.

Expected use case is, depending on the region, full way back allowed or not.
In that context it makes sense to have the way back defined per system to avoid repetition. This behavior can be expanded later on but it'll still be useful to have it per system anyway.

tehhowch commented on source/System.h:
```diff
@@ -164,6 +164,8 @@ class System {
 	// in per frame).
 	double Danger() const;
 
+	void UpdateRandomLinks(System &previousSystem);
+
 
```
Should be a private method.

Hurleveur replied:
if its going to be private it loses its use which is to nitpick what we redefine
so I made it a part of updateneighbors

tehhowch commented on source/Engine.cpp:
```diff
@@ -1238,6 +1238,10 @@ void Engine::EnterSystem()
 	// Remove expired bribes, clearance, and grace periods from past fines.
 	GameData::SetDate(today);
 	GameData::StepEconomy();
+
+	// Refresh random systems that could be linked to this one.
+	const_cast<System *>(system)->UpdateRandomLinks(previousSystem);
+
```
Note that doing this here and just for one system means that the map can be wrong for _every other system._

Hurleveur replied:
I know.
It's done this way for optimization's sake, to not render all the map when we dont need to. That said I could just call the UpdateSystems() method if we don't care about that that much?

Hurleveur replied:
Well I made it redefine the map for all systems at each jump (bye optimization xD)
Only way to make it works without a public updateRandomLinks

tehhowch reviewed (commented):
You should be hooking into the date change system at minimum.

I haven't assessed the sanity of how you are managing links and random links, so the lack of comment there should not be interpreted as tacit approval.

Zitchas commented:
> > Thanks! So, something like `"random link" <system> <chance> <supresswayback>` ?
> 
> There should also be the option to neither guarantee a link is still present immediately after it is traversed nor that that link will be gone.

I chose a poor term. What I meant was "enforce standard link handling", as in ignore the fact that the player arrived using a random link, and strictly use whatever is in the system definition to determine links, if any.

Hurleveur commented:
@Amazinite asked me to change how this works to instead be able to make any element be there or not based on a condition.
I'll think about it a bit, the biggest issue is getting the conditions to the links, since thats all I'll do here. The easiest solution I can see is making PlayerInfo a singleton, or at least having a reference to the conditions somewhere.

Amazinite commented:
What I specifically asked for was to aim for starting work on #7138.
I spoke on stream today about the difficulty of needing to pass the PlayerInfo around more than we'd like just to get access to stuff like the player's conditions. #5415 is an example of an issue that's run into that in the past. If we want something like #7138 where the use of conditions would need to be more wide-spread, we either need some manner of accessing player information or updating what needs that player information that doesn't require passing around PlayerInfo so much.

Hurleveur commented:
Looking at that PR it seems playerinfo were just passed along.
I dont think GameData should own playerinfo but there should be a way to access it indeed.

Here we could just pass along a pointer to the player when updating the systems. Later on it wont require much changes since I assume we'll make the existance of objects part of that update function.

For the syntax I'm thinking you just have to do a subnode, and that will be the conditions of the link.

Hurleveur commented:
So to access the Conditions we have multiple possibilities:
-setting it in UniverseObject at the start in main.cpp (either just the conditions or the player too), or later on by Engine.cpp, or even set it directly in System.cpp
-passing it along like I did here (but I passed the player not just the conditions)
-~~making playerinfo a singleton~~
-making the conditionset somehow accessible from anywhere

Hurleveur commented:
I have no idea why some are failing the test whilst other succeed. Could be due to smth that wasnt initialized properly but I doubt it

Hurleveur commented:
I choose to have the `applicable` implicit, because my answer to these two questions: what else could be defined as a child node, and what else could these conditions mean is nothing else.

Zitchas commented:
> Now, what I meant was a per system attribute; `"guaranteed link back"` which I just added. If you write remove "guaranteed way back" then there wont be one.

So, if I'm understanding this correctly, if, say, Alpha Centauri has
```
link "Sol"
    random < 40
```
and then Sol has:
```
link "Alpha Centauri"
    random < 50
```

When I take the link from Sol to Alpha Centauri and look around, there should be a 40% chance that the link back to Sol is present for me to take, correct?

And then I might have to land and take off a time or two in order to get the link back to sol to show up, this is expected. When it does, and I jump back to Sol, if Sol has this:
```
system Sol
    ...
    "guaranteed link back"
```

Then Sol checks to see if I just arrived via random link, and if I did, it ensures that said link is open to return.

Am I understanding that all correctly?

If I am, then all my previous comments about enforcing standard link handling and whatnot can be ignored, because this above behavior is exactly what I wanted to be default. Namely, that, just like the present, I think everything to do with the link should be controlled by the system the link is going from.

Side questions:
1) Does the "guaranteed link back" check to see where the player came from, and only enforce this link back on the link that the player arrived in? Or does it act like a more general thing and if the player arrived via random link, just force all random links to be working?
2) What happens if a system has this attribute, but doesn't have the corresponding link? Right now we can make one way hyperlanes, and it makes sense to me that in a maze situation we would probably have some one way links. If we had a system "A" that, say, we wanted to guarantee that if they arrived from system "B", they had a guaranteed wayback, but we don't want them to be able to go back to system "C" at all, so while there is a random link coming in from "C", there's no corresponding link out to it. Would this situation work? (In other words, if the player jumps in across the random link from system "B", they will guaranteed have the link back to it for a day; but if they jump in from system "C", then this attribute basically does nothing because "A" doesn't have a link out to "C" to force open.

Hurleveur commented:
Garanteed way back is gone, you'll instead have to use the "previous system: <system>" autocondition that I added in the PR right before this, with "entered system by: <entry>"

Zitchas commented:
Even better! Much tighter control that way.

quyykk commented:
Isn't this basically #7947 but with another implementation?

Hurleveur commented:
not at all, that other PR is for being able to write conditions where we can put numbers
this PR is for using conditions to display objects or not, for now only links

UnorderedSigh commented:
> What I specifically asked for was to aim for starting work on https://github.com/endless-sky/endless-sky/issues/7138 by allowing the existence of links in a system to be based on player conditions. Whatever implementation used here could then potentially be spread to other attributes.

I implemented a framework to easily connect conditions to other things in #4947

Saugia commented:
"You" did, huh? 🤔

UnorderedSigh commented:
> "You" did, huh? thinking

Sorry! Wrong PR number.

UnorderedSigh commented:
It is titled, "Conditions in fleets, and framework for adding conditions elsewhere" because that is what it does. Not only is it implemented, it is also quite fast and has a rather extensive testing framework (1200 lines of new unit tests).

RisingLeaf commented on source/System.cpp:
```diff
@@ -99,7 +100,7 @@ void System::Load(const DataNode &node, Set<Planet> &planets)
 	// For the following keys, if this data node defines a new value for that
 	// key, the old values should be cleared (unless using the "add" keyword).
-	set<string> shouldOverwrite = {"asteroids", "attributes", "belt", "fleet", "link", "object", "hazard"};
+	set<string> shouldOverwrite = {"asteroids", "attributes", "belt", "fleet", "hazard", "link", "object"};
```
why are we changing the order here?

Hurleveur replied:
Alphabetical
Warp asked earlier when I was changing this and I think its fine to keep it?

RisingLeaf replied:
yea fine, was just confused

RisingLeaf reviewed (approved):
Except for this one comment this looks good.
I have not looked at the tests though.

One thing I will note is that I am not quite comfortable with passing player around a lot.

Hurleveur commented:
I hate passing the player as well
I think tjat the player conditions should be accessible globally, but I'd like some input on what ppl think is the best way

Hurleveur commented:
I'm going to be going on a trip for 3 months, meaning I will be much less active on my PRs.
That said, they all are as ready as I could make them, which is why I'm not closing them or marking them as waiting for op.

I won't be able to test and make changes fast, but feel free to post comments and especially to do the needed changes, I truly don't mind. In fact I'd like that.

If this PR isnt the way we want to go with conditional map everything then feel free to close it, but currently any expansion of it will not require changing anything else. Anything in systems can be linked to conditions with this code.

petervdmeer commented:
Fixed a merge-confllict by merging master.

DarcyManoel commented:
Please forgive my haste, but can we also expect conditional objects in the near future? (conditional existence, we already have conditional access to a degree)

Missions can handle system objects, albeit poorly, much like how they could handle system links. Objects are a natural extension of this functionality, as supported by the PR description also.

Hurleveur commented:
I dont think it'd be too hard if one builds on this logic
I dont really have the time for it though, so it'd have to be smbd else

Zitchas commented:
Can we put this on the orphan tag or somehow indicate that this is desired functionality that isn't in the game anymore; or if there is an issue, link to it in here?
